File: leetcode/python/22.GenerateParentheses.py
class Main:

  def generateParenthesis(self,n:int):
    result = []
    self.process("",n,n,result)
    return result
  
  def process(self,prefix,left,right,result):
    if left == 0 and right == 0:
      result.append(prefix)
      return
    if left > 0 :
      self.process(f"{prefix}(",left-1,right,result)
    if left < right:
      self.process(f"{prefix})",left,right-1,result)

sol = Main()
print(sol.generateParenthesis(3))

File: leetcode/js/22.GenerateParentheses.js
class Main{
    generateParenthesis(n){
      var result = []
      this.process("",n,n,result)
      return result
    }
  
    process(prefix,left,right,result){
      if(left===0 && right===0){
        result.push(prefix)
        return
      }
      if(left>0){
        this.process(`${prefix}(`,left-1,right,result)
      }
      if(left<right){
        this.process(`${prefix})`,left,right-1,result)
      }
    }
  }
  
  let sol = new Main()
  console.log(sol.generateParenthesis(3))